pyvast/vast.py

```diff
@@ -1,55 +1,79 @@
 
 """
 
 import asyncio
 import logging
-import pyarrow
-
-
-async def spawn(*args):
-    """Spawns a process asynchronously."""
-    proc = await asyncio.create_subprocess_exec(*args, stdout=asyncio.subprocess.PIPE)
-    return proc
 
 
 class VAST:
     """A VAST node handle"""
 
-    def __init__(self, app="vast", endpoint="localhost:42000"):
+    def __init__(self, binary="vast", endpoint="localhost:42000"):
         self.logger = logging.getLogger("vast")
-        self.app = app
+        self.binary = binary
         self.endpoint = endpoint
-        self.logger.debug("connecting to vast on %s", self.endpoint)
-
-    async def query(self, expression):
-        """Extracts data from VAST according to a query"""
-        self.logger.debug("running query %s", expression)
-        proc = await spawn(self.app, "export", "arrow", expression)
-        # This cannot be avoided, but the reader does not create a second copy,
-        # so we can only do better with memory mapping, i.e. Plasma.
-        output = (await proc.communicate())[0]
-        reader = pyarrow.ipc.open_stream(output)
-        table = reader.read_all()
-        return table
-
-    async def connect(self):
-        """Checks if the endpoint can be connected to"""
-        proc = await spawn(self.app, "status")
-        await proc.communicate()
-        result = False
-        if proc.returncode == 0:
-            result = True
-        return result
+        self.call_stack = []
+        self.logger.debug(f"VAST client configured to use endpoint {self.endpoint}")
+
+    async def __spawn(self, *args, stdin=None):
+        """Spawns a process asynchronously."""
+        return await asyncio.create_subprocess_exec(
+            self.binary,
+            "-e",
+            self.endpoint,
+            *args,
+            stdin=stdin,
+            stdout=asyncio.subprocess.PIPE,
+        )
+
+    async def test_connection(self):
+        """Checks if the configured endpoint is reachable"""
+        proc = await self.__spawn("status")
+        stdout, stderr = await proc.communicate()
+        return proc.returncode == 0
+
+    async def exec(self, stdin=None):
+        self.logger.debug(f"Executing call stack: {self.call_stack}")
+        if not stdin:
+            proc = await self.__spawn(*self.call_stack)
+        else:
+            self.logger.debug(f"Forwarding stdin {stdin}")
+            proc = await self.__spawn(*self.call_stack, stdin=asyncio.subprocess.PIPE)
+            proc.stdin.write(str(stdin).encode())
+            await proc.stdin.drain()
+            proc.stdin.close()
+        self.call_stack = []
+        return await proc.communicate()
+
+    def __getattr__(self, name, **kwargs):
+        """Chains every unknown method call to the internal call stack."""
+        if name.startswith("_"):
+            # trim leading underscores to overcome the 'import' keyword
+            name = name[1:]
+        self.call_stack.append(name)
```

Comment:
We should have a way to check if `name` is allowed in this place. We could use a tree represented as a nested dict for that.
```
commands = {
  "export": {
    "ascii": {},
    "csv": {},
    "json": {},
    "zeek": {}
  },
  "import": {
     ...
  },
  ...
}
```
For example.